packages/cypress/src/add-commands/login.ts

```diff
@@ -0,0 +1,290 @@
+  /**
+   * The user login credentials.
+   */
+  login?: LoginCredentials;
+};
+// Alias for backwards compatibility
+export type CommandLoginByOidcOptions = CommandLoginOptions;
+
+function loginByForm(commandOptions: CommandLoginOptions) {
+  if (isLocalhost()) {
+    throw new Error(
+      `At the moment, the "loginByForm" command only works when testing a Merchant Center production URL. Testing an application running on localhost is not supported due to issues with "cy.origin".`
+    );
+  }
+
+  const projectKey = commandOptions.projectKey ?? Cypress.env('PROJECT_KEY');
+
+  cy.task(
+    'customApplicationConfig',
+    {
+      entryPointUriPath: commandOptions.entryPointUriPath,
+      dotfiles: commandOptions.dotfiles,
+    },
+    // Do not show log, as it may contain sensible information.
+    { log: false }
+  ).then((appConfig: ApplicationRuntimeConfig['env']) => {
+    let url = `/${projectKey}/${commandOptions.entryPointUriPath}`;
+    if (commandOptions.entryPointUriPath === 'account') {
+      url = `/${commandOptions.entryPointUriPath}`;
+    }
+
+    // Log loaded application config for debugging purposes.
+    Cypress.log({
+      displayName: 'task',
+      name: 'customApplicationConfig',
+      message: appConfig,
+    });
+
+    const userCredentials = commandOptions.login ?? {
+      email: Cypress.env('LOGIN_EMAIL') ?? Cypress.env('LOGIN_USER'),
+      password: Cypress.env('LOGIN_PASSWORD'),
+    };
+    const sessionKey = [
+      'loginByForm',
+      userCredentials.email,
+      commandOptions.entryPointUriPath,
+    ];
+    // const mcUrl = new URL(appConfig.mcApiUrl);
+    // const mcFrontendHostname = mcUrl.hostname.replace('mc-api', 'mc');
+
+    function authCallback() {
+      cy.visit(url);
+
+      // NOTE: using `cy.origin` is currently disabled as it does not seem to properly work.
+      // Interestingly, starting an application locally using Vite works, however not when using Webpack.
+      // For now we keep it disabled until we find a solution.
+
+      // // https://cypress.io/blog/2022/04/25/cypress-9-6-0-easily-test-multi-domain-workflows-with-cy-origin/
+      // cy.origin(
+      //   mcFrontendHostname,
+      //   { args: userCredentials },
+      //   fillLoginForm
+      // );
+      fillLoginForm(userCredentials);
+
+      // Wait for the route to be loaded so that the session can be saved.
+      cy.url().should('include', url);
+    }
+
+    // For backwards compatibility.
+    if (Cypress.config('experimentalSessionAndOrigin')) {
+      // https://www.cypress.io/blog/2021/08/04/authenticate-faster-in-tests-cy-session-command/
+      cy.session(sessionKey, authCallback, {
+        validate() {
+          validateUserSession(appConfig.mcApiUrl);
+        },
+      });
+    } else {
+      cy.log(
+        `We recommend turning on the flag "experimentalSessionAndOrigin" to be able to use "cy.session" and thus reduce the time to log in between tests.`
+      );
+      authCallback();
```

Comment:
**nit**: I was wondering why we don't validate the user session in this code branch. Is it for performance reasons? (we don't want to validate the session on every login of every test when we're not leveraging session caching)

Reply:
You mean to manually call `validateUserSession` part?

I haven't thought about it, good point. In theory validating the session is not strictly required. I assume the tests will fail anyway if there is no session.

Do you think it would make sense to do it here as well?

Reply:
If tests will fail anyway when login did not succeed, I guess I would remove that validation from the other code branch.
What was confusing for me is that we are only running the validating when using the cached session. I see this is done to ensure the reused session is still valid but it seems the test will fail in that case.